### PR TITLE
Use opaque types for handle references

### PR DESCRIPTION
**Pull request by RReverser**
This improves typechecking by verifying that user passes pointers of correct types.

### CODE REVIEW

File: src/binaryen-c.h [hunk 1]
```diff
@@ -64,6 +64,16 @@
 #define BINARYEN_API
 #endif
 
+#ifdef __cplusplus
+#define BINARYEN_REF(NAME)                                                     \
```
Comment (RReverser) [inline src/binaryen-c.h]: @kripken You added your review before I added this macro, so wanted to double-check if this is okay or, if not, can you suggest a better name?

Comment (kripken) [reply src/binaryen-c.h]: I don't have a better idea, lgtm.

**Comment by RReverser** [comment]
Looks like CI is failing here for the same reasons as on master (on macOS only), but otherwise should be good to go.

@kripken can you merge please? (I don't have rights on this repo/org)

**Comment by kripken** [comment]
I'm trying to fix CI now in https://github.com/WebAssembly/binaryen/pull/2474 - after that lands, we can merge it here, then land. (Unless we are sure this can land without CI testing? That's not obvious to me.)

**Comment by kripken** [comment]
That other PR landed.

**Comment by RReverser** [comment]
> Unless we are sure this can land without CI testing? That's not obvious to me.

I was ~sure only because I checked Travis logs and saw that all other builders passed tests successfully, and only macOS one failed to start. But fixing the final builder is even better indeed :)

**Comment by kripken** [comment]
Great, thanks @RReverser!

**Pull request merged**